torch/_functorch/aot_autograd.py

```diff
@@ -1392,6 +1395,20 @@ def view_avoid_dupes_with_primals(t):
             f_fw_graph_outs = f_fw_graph_outs + intermediate_bases
         fw_graph_outs = pytree.tree_map(from_fun, f_fw_graph_outs)
 
+        set_grad_enabled = None
+        if ctx := torch._guards.TracingContext.get():
+            prior_grad_enabled = ctx.global_context.global_state.get('grad_enabled')
```

Comment:
This isn't the right way to get this information. In particular, we won't generate the correct grad mode mutation if you call aot autograd by itself, without an enclosing Dynamo context.

The most appropriate way to get this information is to detect it when we actually run the graph. run_functionalized_fw_and_collect_metadata is not a bad place to put it, because this is the first place we run the graph in question, and we're trying to get metadata about tensor mutation at this point in time.

There are some extra things that should be tested too. For example, once we've run the graph once, do we reset the grad enabled state to what it was previously before we run the graph again? If we don't, we could be generating more/less autograd nodes than is necessary. I don't remember ever reviewing code that explicitly handles this in aot autograd, so proving to yourself that this actually works would also be useful.

Reply:
> we won't generate the correct grad mode mutation if you call aot autograd by itself, without an enclosing Dynamo context.

May I know what current users of this would be?

Reply:
> The most appropriate way to get this information is to detect it when we actually run the graph.

Unfortunately, unlike with tensors for which we have example inputs, in the grad mode case we do not have a proxy by which we can trace mutations.

That is, unless we can dispatch the call to set_grad_mode so as to increment a mutation counter, from which we can detect mutations and simply set the mutation epilogue to the last set value.

Reply:
Furthermore, that is assuming the grad mode value is a constant; for instance, we might have the following data-dependent value.

Hence, not assuming the const-ness of the ambient grad mode is likely to fail.

```python
torch.set_grad_enabled(not torch.is_grad_enabled())

Reply:
@jon-chuang why not just directly check `torch.is_grad_enabled()`  right before/after the first time we execute the torch graph? That happens here: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1059

> Hence, not assuming the const-ness of the ambient grad mode is likely to fail.

As long as we reset the grad mode to the "correct value" right before running the function, then I think this should work out: the code snippet you put above will just grab the current ambient grad state, and it's on us to make sure it's set to whatever the ambient grad state would have been in eager mode.

So in particular: right before we trace out the joint graph (one option would be [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1709)), we will need to make sure to set the ambient grad state to whatever it was supposed to be.

Reply:
> why not just directly check torch.is_grad_enabled() right before/after the first time we execute the torch graph? 

That's because Dynamo itself will apply the mutation and not reset it. So the grad value at the time we are handed the graph is not the grad value prior to capturing the graph.

Reply:
In general, we cannot be assured about how the graph handed to aotautograd was captured and what the ambient values were at the time.

Reply:
Perhaps what we need is for dynamo to reset the ambient states prior to calling the backends? 🤔

Reply:
> https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1059

Is there any other location where we run a graph with mutating side-effects inside of aot_autograd or other backends?

Reply:
> That's because Dynamo itself will apply the mutation and not reset it. So the grad mode value at the time we are handed the graph is not the grad mode value prior to capturing the graph.

Dynamo needs to reset it lol. Sorry, this is kind of a rats nest right now. And not just grad enabled, all of the side effects that we are using the "real" state to modify them need to get reset as well.

Reply:
>> Is there any other location where we run a graph with mutating side-effects inside of aot_autograd or other backends?
>
> In particular, is run_functionalized_fw_and_collect_metadata the only place where we do so inside of aot_autograd?

This question is particularly relevant if there is anything that depends on ambient grad mode before we call aotautograd.

Reply:
> And not just grad enabled, all of the side effects that we are using the "real" state to modify them need to get reset as well.

Well, this PR resets all of the autocasting/torch_function_enabled for dynamo, but not for aotautograd.

autocasting/torch_function_enabled are not as problematic as the user-facing APIs encourage ctx manager, and dynamo ensures that it exits all in-graph ctx managers prior to calling the user compiler.

Reply:
Note that the autocasting part is still feeble due to https://github.com/pytorch/pytorch/issues/112260, but once we fix that it should be pretty solid.